dapps/src/handlers/fetch.rs

```diff
@@ -296,100 +316,54 @@ impl<H: ContentValidator, F: Fetch> ContentFetcherHandler<H, F> {
 					warn!(target: "dapps", "Unable to fetch content: {:?}", e);
 					FetchState::Error(errors.download_error(e))
 				},
-			};
-			// Content may be resolved when the connection is already dropped.
-			let _ = tx2.send(new_state);
-			// Ignoring control errors
-			let _ = control.ready(Next::read());
-			Ok(()) as Result<(), ()>
+			})
 		});
 
 		// make sure to run within fetch thread pool.
-		let future = self.fetch.process(future);
-		// spawn to event loop
-		let control = self.control.clone();
-		let errors = self.errors.clone();
-		self.remote.spawn_with_timeout(|| future, Duration::from_secs(FETCH_TIMEOUT), move || {
-			// Notify about the timeout
-			let _ = tx.send(FetchState::Error(errors.timeout_error()));
-			// Ignoring control errors
-			let _ = control.ready(Next::read());
-		});
-
-		rx
+		fetch.process(future)
 	}
 }
 
-impl<H: ContentValidator, F: Fetch> server::Handler<HttpStream> for ContentFetcherHandler<H, F> {
-	fn on_request(&mut self, request: server::Request<HttpStream>) -> Next {
-		let status = if let FetchState::NotStarted(ref url) = self.status {
-			let uri = request.uri().clone();
-			let installer = self.installer.take().expect("Installer always set initialy; installer used only in on_request; on_request invoked only once; qed");
-
-			Some(match *request.method() {
-				// Start fetching content
-				Method::Get => {
-					trace!(target: "dapps", "Fetching content from: {:?}", url);
-					let receiver = self.fetch_content(uri, url, installer);
-					FetchState::InProgress(receiver)
-				},
-				// or return error
-				_ => FetchState::Error(self.errors.method_not_allowed()),
-			})
-		} else { None };
-
-		if let Some(status) = status {
-			self.fetch_control.set_status(&status);
-			self.status = status;
-		}
+impl Future for ContentFetcherHandler {
+	type Item = hyper::Response;
+	type Error = hyper::Error;
 
-		Next::read()
-	}
+	fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
+		loop {
+			trace!(target: "dapps", "Polling status: {:?}", self.status);
+			self.status = match mem::replace(&mut self.status, FetchState::Empty) {
+				FetchState::Error(error) => {
+					return Ok(futures::Async::Ready(error.into()));
+				},
+				FetchState::Streaming(response) => {
+					return Ok(futures::Async::Ready(response));
+				},
+				any => any,
```

Comment:
imo, it's better to write: 

```rust
if let FetchState::Error(ref mut error) = self.status {
  // here do mem::replace on error
  // and return Ok(..)
}

if let FetchState::Streaming(ref mut response) = self.response {
  // here do mem::replace on response
  // and return Ok(..)
}

```

Reply:
That doesn't give you ownership of `error` required for `return`.